Allow commit txs with multiple UTxO

ch1bo opened the pull request:
:1234: Extends the specification and the on-chain protocol to record any number of outputs committed in a commit tx into a Hydra head.

:1234: Changes (and simplifies) multiple call sites from `Maybe Commit` to `[Commit]`.

:1234: Adds a mutation test where a commit is tried to be left out.

:1234: This _theoretically_ increases the likelihood that some party locks up too many UTxO and makes the head not even abortable (see #699). Practically, this issue already exists with reference scripts and big datums.

:1234: Removes the `MoreThanOneUTxOCommitted` error as it's not needed anymore.

--- 

TODO:
- [ ] Update the spec / alternative version
  - [x] Requires #693
  - [x] Do the update
  - [ ] Update the commitTx picture (needed?)
- [x] Think about the fact that this makes the head not abortable more likely (see #699)
  - While it's a bit simpler to shoot yourself in the foot, the risk is already there. Let's not worry and rather fix it with #699 or so.
- [x] Add/update tests which leverage the fact that there are now more than one committed UTxO (new mutations?)

<!-- Tick off or strike-through / remove if not applicable -->
* [x] CHANGELOG updated
* [x] Documentation updated
* [x] Added and/or updated haddocks
* [x] No new TODOs introduced or explained herafter

ch1bo committed:
Allow multiple commits

This changes the Redeemer and Datum types of commitTx to be [TxOutRef]
instead of Maybe TxOutRef and generalize the checks on- and off-chain to
work on a list of committed UTxO instead of an optional, single one.

ch1bo committed:
Update tx-cost benchmark to show cost of commit

Now that we allow multiple commits, we should improve the benchmark to
show the (sad) cost of this.

ch1bo committed:
Support multiple committed UTxO in tests

This change is mostly fixing compilation issues. Not all tests are
testing the fact that there now may be more than one commit.

ch1bo committed:
Remove MoreThanOneUTxOCommitted error

This also requires generating new golden files which had this error
included.

ffakenz commented on hydra-cluster/test/Test/DirectChainSpec.hs:
```diff
@@ -187,13 +187,8 @@ spec = around showLogsOnFailure $ do
-            someUTxOB <- generate $ genOneUTxOFor aliceCardanoVk
-
-            postTx (CommitTx alice (someUTxOA <> someUTxOB))
-              `shouldThrow` (== MoreThanOneUTxOCommitted @Tx)
```
should not we have a spec checking this is now valid?

ch1bo replied:
We have a valid commit transaction interaction just below on line 201/202

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/TxSpec.hs:
```diff
@@ -187,15 +187,14 @@ generateCommitUTxOs parties = do
   txins <- vectorOf (length parties) (arbitrary @TxIn)
   let vks = (\p -> (genVerificationKey `genForParty` p, p)) <$> parties
   committedUTxO <-
```
should we use the plural to name this variable?

ch1bo replied:
There has been discussions of what the plural of `UTxO` is and some responses are "it's already plural" :sweat_smile:

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/Contract/CollectCom.hs:
```diff
@@ -144,7 +144,7 @@ data HealthyCommit = HealthyCommit
 healthyCommitOutput ::
   Party ->
-  (TxIn, TxOut CtxUTxO) ->
+  UTxO ->
```
should this be a list?

ch1bo replied:
UTxO may already contain multiple entries. What name would be more intuitive? `UTxOSet` or `UTxOMap`?

ffakenz replied:
UTxO is fine!

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/Contract/CollectCom.hs:
```diff
@@ -93,7 +93,7 @@ healthyCommits =
  where
   committedUTxO =
     generateWith
-      (replicateM (length healthyParties) genCommittableTxOut)
```
should `genCommittableTxOut` not produce only ada utxo?

ch1bo replied:
Yeah that is something we could incorporate in this test suite as well. I was just porting what was there before.

ffakenz commented on hydra-plutus/src/Hydra/Contract/Initial.hs:
```diff
@@ -125,15 +127,17 @@ checkCommit commitValidator headId committedRef context =
+  committedUTxO = do
+    flip fmap committedRefs $ \ref ->
+      case findTxInByTxOutRef ref txInfo of
+        Nothing -> traceError "outref not found"
```
I guess we need a new error code for this

ch1bo replied:
Done in https://github.com/input-output-hk/hydra/commit/116081f9da4ec8179005b675613197350a8f14c8

ffakenz commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -170,9 +170,8 @@ commitTx ::
-  -- We currently limit committing one UTxO to the head because of size limitations.
-  Maybe (TxIn, TxOut CtxUTxO) ->
+  -- | The UTxO to commit to the Head
+  UTxO ->
```
why this is not a list?

ffakenz replied:
nvm

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/Contract/Commit.hs:
```diff
@@ -141,7 +142,7 @@ genCommitMutation (tx, _utxo) =
   commitTxOut = fromJust $ txOuts' tx !!? 0
 
-  (committedTxIn, committedTxOut) = healthyCommittedUTxO
+  (committedTxIn, committedTxOut) = List.head $ UTxO.pairs healthyCommittedUTxO
```
should we not produce many instead of a single one?

ffakenz commented on hydra-node/exe/tx-cost/TxCost.hs:
```diff
@@ -82,8 +82,8 @@ computeInitCost = do
-  interesting <- catMaybes <$> mapM compute [1, 2, 3, 5, 10, 50, 100]
-  limit <- maybeToList . getFirst <$> foldMapM (fmap First . compute) [500, 499 .. 101]
+  interesting <- catMaybes <$> mapM compute [1, 2, 3, 5, 10]
+  limit <- maybeToList . getFirst <$> foldMapM (fmap First . compute) [100, 99 .. 11]
```
were these values got reduced due to changes in the scripts?

why u say the cost of this is sad now?

ch1bo replied:
We were not actually testing anything else than `1`. Sad because the maximum seems to be around `13` ada-only UTxO`, i.e. it's very expensive.

ffakenz reviewed (commented):
This looks very good! 💯 

What is left for me is:
- a test to validate now multiple commits are allowed.
- have a mutation to validate if any of the multiple commits is not healthy.
- update api documentation to mention now we support multiple commits.
- maybe explain and mention the limits around this.

Also what is not clear to me is how, from API can I commit multiple UTxO.
Seems the user must provide a big one instead of multiple, is that intended?

ch1bo committed:
Update golden scripts in hydra-plutus

ch1bo committed:
Reflect changes of multiple commits in specification

The picture for commitTx was not yet updated in this.

GeorgeFlerovsky commented:
This PR works well for us at `hydra-auction`. Very helpful and no issues. :pray:

ch1bo committed:
Update known issues

The limit of comitted UTxO is not in place anymore. Also, reorder two
bullets to group known issues related to commits.

ch1bo committed:
Update changelog and bump hydra-plutus version

ch1bo committed:
Make CollectCom mutation tests use a varying number of committed UTxO

Still ada-only (for better control of size), but within a range of 0-5
outputs are committed.

The scenario is static per compilation (constant seed, but depends on
generators), but currently UTxO with lengths of [0,2,4] is collected
from three parties.

ch1bo committed:
Make Commit mutation test use multiple healthyCommitUTxO

Often we only modify a single one of them.

ch1bo committed:
Add mutation to leave out a claimed to commit UTxO

This is a bit artifical, but ensures we check for completeness of
recorded commits.

ch1bo committed:
Add an error code for commit tx not finding outref
